packages/viewer/src/stream/dispatcher.ts

```diff
@@ -0,0 +1,70 @@
+import { Listener, Disposable, Async } from '@vertexvis/utils';
+import { StreamApi } from '@vertexvis/stream-api';
+import { vertexvis } from '@vertexvis/frame-streaming-protos';
+
+export const DEFAULT_TIMEOUT_IN_MS = 1000 * 10;
+
+export class StreamApiEventDispatcher<T> {
+  private listeners: Array<Listener<T>> = [];
+  private eventDisposable?: Disposable;
+  private requestDisposable?: Disposable;
+
+  public constructor(
+    private stream: StreamApi,
+    private predicate: (
+      msg: vertexvis.protobuf.stream.IStreamMessage
+    ) => boolean,
+    private transform: (msg: vertexvis.protobuf.stream.IStreamMessage) => T,
+    private timeout: number = DEFAULT_TIMEOUT_IN_MS
+  ) {
+    this.handleMessage = this.handleMessage.bind(this);
+  }
+
+  public on(listener: Listener<T>): void {
+    this.listeners = [...this.listeners, listener];
+    if (this.eventDisposable == null || this.requestDisposable == null) {
+      this.addListeners();
+    }
+  }
+
+  public off(listener: Listener<T>): void {
+    this.listeners = this.listeners.filter((l) => l !== listener);
+    if (this.listeners.length === 0) {
+      this.removeListeners();
+    }
+  }
+
+  public once(listener?: Listener<T>): Promise<T> {
```

Comment:
What's the purpose of this listener? Can't the caller just do `.then(listener)`?

Reply:
Yeah, this probably isn't necessary -- I'll remove